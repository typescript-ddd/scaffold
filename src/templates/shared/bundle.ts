import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
