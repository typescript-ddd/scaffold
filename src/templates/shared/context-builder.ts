import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
