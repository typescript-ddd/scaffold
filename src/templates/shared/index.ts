export * from "./context-builder";
export * from "./generate-context";