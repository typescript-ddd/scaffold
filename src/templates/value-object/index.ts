export * from "./value-object.generator";
export * from "./value-object.template";
export * from "./value-object.types";
