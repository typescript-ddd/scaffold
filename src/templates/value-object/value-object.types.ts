import {
  TemplateBooleanOption,
  TemplatePropertiesOption,
  TemplatePropertyValue,
  TemplateStringOption,
} from "../shared/template";

export type ValueObjectTemplateValues = {
  name: string;
  valueType: string;
  valueProps?: TemplatePropertyValue[];
  validationErrorMessage?: string;
  generateValueGetters?: boolean;
};

export type ValueObjectTemplateOptions = {
  name: TemplateStringOption;
  valueType: TemplateStringOption;
  valueProps?: TemplatePropertiesOption;
  validationErrorMessage?: TemplateStringOption;
  generateValueGetters?: TemplateBooleanOption;
};
