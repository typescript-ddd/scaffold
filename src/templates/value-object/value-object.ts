import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateValueObjectOptions = {
  name: string;
  valueType: string;
  valueProps?: { name: string; type: string }[];
  validationErrorMessage?: string;
  generateValueGetters?: boolean;
};

export const generateValueObject = (
  options: GenerateValueObjectOptions,
  context: GenerateContext
) => {
  const {
    name,
    valueType = "string",
    valueProps,
    validationErrorMessage,
    generateValueGetters = false,
  } = options;
  const className = strings.capitalize(name);

  const project = new Project();
  const sourceFile = project.createSourceFile(
    context.fileName,
    `
export class ${className} extends ValueObject<${valueType}> {}    
`
  );
  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["InvariantViolationError", "ValueObject"],
  });

  context.writeSourceFileImports(sourceFile);

  if (valueProps?.length) {
    sourceFile
      .insertInterface(1, {
        name: valueType,
        properties: valueProps.map((prop) => ({
          name: prop.name,
          type: prop.type,
        })),
        isExported: true,
      })
      .addJsDoc({
        description: `Represents the value properties of ${strings.describe(
          className
        )}.`,
      });
  }

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(name)} value.`,
    tags: [{ tagName: "extends", text: "{ValueObject}" }],
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: valueType }],
  });

  constructorDeclaration.setBodyText((writer) => {
    writer.write("super(value);");
    writer.writeLine(`if (!${className}.validate(value)) {`);
    writer.writeLine(
      `  throw new InvariantViolationError("${
        validationErrorMessage ||
        `Invalid value for ${strings.describe(className)}.`
      }");`
    );
    writer.writeLine(`}`);
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} value.`,
    tags: [
      {
        tagName: "param",
        text: `{${valueType}} value - ${strings.describe(valueType)} value.`,
      },
    ],
  });

  if (valueProps?.length && generateValueGetters) {
    valueProps.forEach((prop) => {
      classDeclaration
        .addMethod({
          name: `get${strings.capitalize(prop.name)}`,
          returnType: prop.type,
        })
        .setBodyText(`return this.value.${prop.name};`)
        .insertJsDoc(0, {
          description: `Gets the ${strings.lower(prop.name)} value.`,
          tags: [
            {
              tagName: "returns",
              text: `{${prop.type}} - The ${strings.lower(prop.name)} value.`,
            },
          ],
        });
    });
  }

  classDeclaration
    .addMethod({
      name: "selfEquals",
      returnType: "boolean",
      scope: Scope.Protected,
      parameters: [{ name: "other", type: className }],
    })
    .setBodyText((writer) => {
      if (["string", "number"].includes(valueType)) {
        writer.writeLine("return this.value === other.value;");
        return;
      }
      writer.writeLine("//TODO: Implement value comparison");
      writer.writeLine("return false;");
    })
    .insertJsDoc(0, {
      tags: [{ tagName: "inhertiDoc" }],
    });

  classDeclaration
    .addMethod({
      name: "validate",
      isStatic: true,
      parameters: [{ name: "value", type: valueType }],
      returnType: "boolean",
    })
    .setBodyText((writer) => {
      writer.writeLine("// TODO : Implement validation");
      writer.writeLine("return true;");
    })
    .insertJsDoc(0, {
      description: `Validates the value of ${strings.describe(name)}.`,
      tags: [
        {
          tagName: "param",
          text: `{${valueType}} value - The value to validate.`,
        },
        {
          tagName: "returns",
          text: `{boolean} - True if the value is valid, otherwise false.`,
        },
      ],
    });

  classDeclaration
    .addMethod({
      name: "create",
      isStatic: true,
      parameters: [{ name: "value", type: "string" }],
      returnType: className,
    })
    .setBodyText(`return new ${className}(value);`)
    .insertJsDoc(0, {
      description: `Creates a new instance of ${strings.describe(name)} value.`,
      tags: [
        {
          tagName: "returns",
          text: `{${className}} - ${strings.capitalize(
            strings.describe(name)
          )} value.`,
        },
      ],
    });

  let output = sourceFile.getFullText();

  return output;
};
