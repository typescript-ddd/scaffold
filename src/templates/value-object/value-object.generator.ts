import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { Chunk, GenerateContext } from "../shared";
import { ValueObjectTemplateValues } from "./value-object.types";

export const generateValueObject = (
  values: ValueObjectTemplateValues,
  context: GenerateContext,
  chunkName?: string
): Chunk => {
  const {
    name,
    valueType = "string",
    valueProps,
    validationErrorMessage,
    generateValueGetters = false,
  } = values;
  const className = strings.capitalize(name);
  const {
    fileName = `${strings.kebab(name)}.ts`,
    projectPath = ["domain", "model"],
  } = context.currentFile || {};
  const project = new Project();
  const sourceFile = project.createSourceFile(
    fileName,
    `
export class ${className} extends ValueObject<${valueType}> {}    
`
  );
  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["InvariantViolationError", "ValueObject"],
  });

  context.writeSourceFileImports(sourceFile);

  if (valueProps?.length) {
    sourceFile
      .insertInterface(1, {
        name: valueType,
        properties: valueProps.map((prop) => ({
          name: prop.name,
          type: prop.valueType,
        })),
        isExported: true,
      })
      .addJsDoc({
        description: `Represents the value properties of ${strings.describe(
          className
        )}.`,
      });
  }

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(name)} value.`,
    tags: [{ tagName: "extends", text: "{ValueObject}" }],
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: valueType }],
  });

  constructorDeclaration.setBodyText((writer) => {
    writer.write("super(value);");
    writer.writeLine(`if (!${className}.validate(value)) {`);
    writer.writeLine(
      `  throw new InvariantViolationError("${
        validationErrorMessage ||
        `Invalid value for ${strings.describe(className)}.`
      }");`
    );
    writer.writeLine(`}`);
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} value.`,
    tags: [
      {
        tagName: "param",
        text: `{${valueType}} value - ${strings.describe(valueType)} value.`,
      },
    ],
  });

  if (valueProps?.length && generateValueGetters) {
    valueProps.forEach((prop) => {
      classDeclaration
        .addMethod({
          name: `get${strings.capitalize(prop.name)}`,
          returnType: prop.valueType,
        })
        .setBodyText(`return this.value.${prop.name};`)
        .insertJsDoc(0, {
          description: `Gets the ${strings.lower(prop.name)} value.`,
          tags: [
            {
              tagName: "returns",
              text: `{${prop.valueType}} - The ${strings.lower(
                prop.name
              )} value.`,
            },
          ],
        });
    });
  }

  classDeclaration
    .addMethod({
      name: "selfEquals",
      returnType: "boolean",
      scope: Scope.Protected,
      parameters: [{ name: "other", type: className }],
    })
    .setBodyText((writer) => {
      if (["string", "number"].includes(valueType)) {
        writer.writeLine("return this.value === other.value;");
        return;
      }
      writer.writeLine("//TODO: Implement value comparison");
      writer.writeLine("return false;");
    })
    .insertJsDoc(0, {
      tags: [{ tagName: "inhertiDoc" }],
    });

  classDeclaration
    .addMethod({
      name: "validate",
      isStatic: true,
      parameters: [{ name: "value", type: valueType }],
      returnType: "boolean",
    })
    .setBodyText((writer) => {
      writer.writeLine("// TODO : Implement validation");
      writer.writeLine("return true;");
    })
    .insertJsDoc(0, {
      description: `Validates the value of ${strings.describe(name)}.`,
      tags: [
        {
          tagName: "param",
          text: `{${valueType}} value - The value to validate.`,
        },
        {
          tagName: "returns",
          text: `{boolean} - True if the value is valid, otherwise false.`,
        },
      ],
    });

  classDeclaration
    .addMethod({
      name: "create",
      isStatic: true,
      parameters: [{ name: "value", type: "string" }],
      returnType: className,
    })
    .setBodyText(`return new ${className}(value);`)
    .insertJsDoc(0, {
      description: `Creates a new instance of ${strings.describe(name)} value.`,
      tags: [
        {
          tagName: "returns",
          text: `{${className}} - ${strings.capitalize(
            strings.describe(name)
          )} value.`,
        },
      ],
    });

  return {
    name: chunkName || "ValueObject",
    fileName,
    projectPath,
    content: sourceFile.getFullText(),
    values,
  };
};
