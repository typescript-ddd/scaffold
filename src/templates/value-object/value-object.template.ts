import { GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateValueObject } from "./value-object.generator";
import {
  ValueObjectTemplateOptions,
  ValueObjectTemplateValues,
} from "./value-object.types";

export class ValueObjectTemplate
  implements Template<ValueObjectTemplateValues, ValueObjectTemplateOptions>
{
  readonly name = "ValueObject";
  readonly description = "Generates a value object";
  readonly options: ValueObjectTemplateOptions = {
    name: {
      name: "name",
      label: "Name",
      description: "The name of the value object",
      optionType: "string",
    },
    valueType: {
      name: "valueType",
      label: "Value Type",
      description: "The value type of the value object",
      optionType: "string",
    },
    valueProps: {
      name: "valueProps",
      label: "Value Props",
      description: "The value props of the value object",
      optionType: "properties",
    },
    validationErrorMessage: {
      name: "validationErrorMessage",
      label: "Validation Error Message",
      description: "The validation error message of the value object",
      optionType: "string",
    },
    generateValueGetters: {
      name: "generateValueGetters",
      label: "Generate Value Getters",
      description: "Whether to generate value getters",
      optionType: "boolean",
    },
  };
  readonly defaultValues: ValueObjectTemplateValues = {
    name: "Phone",
    valueType: "PhoneProps",
    valueProps: [{ name: "number", valueType: "string" }],
    generateValueGetters: true,
    validationErrorMessage: "Invalid phone number",
  };
  generate(values: ValueObjectTemplateValues, context: GenerateContext): string {
    return generateValueObject(values, context);
  }
}
