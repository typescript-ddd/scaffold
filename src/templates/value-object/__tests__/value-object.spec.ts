import { GenerateContext, ContextBuilder } from "../../shared";
import { generateValueObject } from "../value-object";

describe("ValueObject", () => {
  let context: GenerateContext;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
  });

  it("Should render the value object component", () => {
        
    const output = generateValueObject(
      { name: "FirstName", valueType: "string" },
      context
    );

    expect(output).toMatchSnapshot();
  });

  it("Should render the value object component with properties", () => {
    const output = generateValueObject(
      { name: "Phone", valueType: "PhoneProps", valueProps: [{ name: "number", type: "string" }], generateValueGetters: true },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
