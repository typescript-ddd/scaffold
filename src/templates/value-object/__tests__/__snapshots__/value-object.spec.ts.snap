// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueObject Should render the value object component 1`] = `
"import { InvariantViolationError, ValueObject } from "@typescript-ddd/core";

/**
 * Represents a first name value.
 * @extends {ValueObject}
 */
export class FirstName extends ValueObject<string> {
    /**
     * Initializes a new instance of a first name value.
     * @param {string} value - a string value.
     */
    private constructor(value: string) {
        super(value);
        if (!FirstName.validate(value)) {
          throw new InvariantViolationError("Invalid value for a first name.");
        }
    }

    /** @inhertiDoc */
    protected selfEquals(other: FirstName): boolean {
        return this.value === other.value;
    }

    /**
     * Validates the value of a first name.
     * @param {string} value - The value to validate.
     * @returns {boolean} - True if the value is valid, otherwise false.
     */
    static validate(value: string): boolean {
        // TODO : Implement validation
        return true;
    }

    /**
     * Creates a new instance of a first name value.
     * @returns {FirstName} - A first name value.
     */
    static create(value: string): FirstName {
        return new FirstName(value);
    }
}    
"
`;

exports[`ValueObject Should render the value object component with properties 1`] = `
"import { InvariantViolationError, ValueObject } from "@typescript-ddd/core";

/** Represents the value properties of a phone. */
export interface PhoneProps {
    number: string;
}

/**
 * Represents a phone value.
 * @extends {ValueObject}
 */
export class Phone extends ValueObject<PhoneProps> {
    /**
     * Initializes a new instance of a phone value.
     * @param {PhoneProps} value - a phone props value.
     */
    private constructor(value: PhoneProps) {
        super(value);
        if (!Phone.validate(value)) {
          throw new InvariantViolationError("Invalid value for a phone.");
        }
    }

    /**
     * Gets the number value.
     * @returns {string} - The number value.
     */
    getNumber(): string {
        return this.value.number;
    }

    /** @inhertiDoc */
    protected selfEquals(other: Phone): boolean {
        //TODO: Implement value comparison
        return false;
    }

    /**
     * Validates the value of a phone.
     * @param {PhoneProps} value - The value to validate.
     * @returns {boolean} - True if the value is valid, otherwise false.
     */
    static validate(value: PhoneProps): boolean {
        // TODO : Implement validation
        return true;
    }

    /**
     * Creates a new instance of a phone value.
     * @returns {Phone} - A phone value.
     */
    static create(value: string): Phone {
        return new Phone(value);
    }
}    
"
`;
